Using a switch statement together with if- or for-statement without using {} crashes

Using a switch-statement with an if- or for-statement without using {}, the GWT compilation will throw an exception. Surrounding the switch with {} will fix that exception.

```
if ((vpos != null))
    switch (vpos) {
     case TOP :
         return "text-before-edge";
     case CENTER :
         return "central";
     case BASELINE :
         return "baseline";
     case BOTTOM :
         return "text-after-edge";
     }
```
will crash with this exception:

```
[INFO] Caused by: java.lang.ClassCastException: class com.google.gwt.dev.jjs.ast.JSwitchStatement cannot be cast to class com.google.gwt.dev.jjs.ast.JExpression (com.google.gwt.dev.jjs.ast.JSwitchStatement and com.google.gwt.dev.jjs.ast.JExpression are in unnamed module of loader 'app')
[INFO] 	at com.google.gwt.dev.jjs.impl.GwtAstBuilder$AstVisitor.pop(GwtAstBuilder.java:2816)
[INFO] 	at com.google.gwt.dev.jjs.impl.GwtAstBuilder$AstVisitor.endVisit(GwtAstBuilder.java:1087)
[INFO] 	... 25 more
[INFO]    [ERROR] at SvgNodePeer.java(201): if ((vpos != null))
[INFO]     switch (vpos) {
[INFO]     case TOP :
[INFO]         return "text-before-edge";
[INFO]     case CENTER :
[INFO]         return "central";
[INFO]     case BASELINE :
[INFO]         return "baseline";
[INFO]     case BOTTOM :
[INFO]         return "text-after-edge";
[INFO]     }
[INFO]       org.eclipse.jdt.internal.compiler.ast.IfStatement
```

while this will work: 

```
if ((vpos != null)) {
    switch (vpos) {
     case TOP :
         return "text-before-edge";
     case CENTER :
         return "central";
     case BASELINE :
         return "baseline";
     case BOTTOM :
         return "text-after-edge";
     }
}
```

Same issue with this code:

```
                for (InMemoryAuthorizationRule rule : rules)
                    switch (rule.computeRuleResult(operationRequest)) {
                        case DENIED:  result = AuthorizationRuleResult.DENIED; break; // Breaking as it's a final decision
                        case GRANTED: result = AuthorizationRuleResult.GRANTED; // Not breaking, as we need to check there is not another denying rule (which is priority)
                        case OUT_OF_RULE_CONTEXT: // just ignoring it and looping to the next
                    }
```


**GWT version:** 2.12.0
**Browser (with version):** all browser 
**Operating System:** all OS

---

--- a/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
+++ b/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
@@ -2812,26 +2812,31 @@ public class GwtAstBuilder {
 
     protected JStatement pop(Statement x) {
       JNode pop = (x == null) ? null : pop();
-      if (x instanceof Expression) {
+      if (x instanceof Expression && pop instanceof JExpression) {
         return maybeBoxOrUnbox((JExpression) pop, (Expression) x).makeStatement();
       }
       return (JStatement) pop;
     }
 
-    @SuppressWarnings("unchecked")
     protected <T extends JStatement> List<T> pop(Statement[] statements) {
       if (statements == null) {
         return Collections.emptyList();
       }
-      List<T> result = (List<T>) popList(statements.length);
+      List<T> result = Lists.newArrayList();
+      List<? extends JNode> stack = popList(statements.length);
       int i = 0;
-      for (ListIterator<T> it = result.listIterator(); it.hasNext(); ++i) {
+      for (ListIterator<? extends JNode> it = stack.listIterator(); it.hasNext(); ++i) {
         Object element = it.next();
-        if (element == null) {
-          it.remove();
-        } else if (element instanceof JExpression) {
-          it.set((T)
-              maybeBoxOrUnbox((JExpression) element, (Expression) statements[i]).makeStatement());
+        if (element != null) {
+          if (element instanceof JExpression) {
+            JExpression unboxed = maybeBoxOrUnbox((JExpression) element, (Expression) statements[i]);
+            result.add((T) unboxed.makeStatement());
+          } else if (element instanceof JStatement) {
+            result.add((T) element);
+          } else {
+            throw new IllegalStateException(
+                "Unexpected element type, expected statement or expression: " + element);
+          }
         }
       }
       return result;

--- a/dev/core/test/com/google/gwt/dev/jjs/impl/Java17AstTest.java
+++ b/dev/core/test/com/google/gwt/dev/jjs/impl/Java17AstTest.java
@@ -397,6 +397,65 @@ public class Java17AstTest extends FullCompileTestBase {
           "}");
   }
 
+  // The JDT ast nodes for both switch statements and expressions extend Expression, and specific
+  // ast builder traversals previously. Test switch expressions/statements where statements/blocks
+  // can be encountered.
+  public void testSwitchesWithoutBlocks() throws UnableToCompleteException {
+    compileSnippet("void",
+        "for (int i = 0; i < 10; i++) " +
+            "  switch(i) { " +
+            "    case 1: break;" +
+            "    case 2: " +
+            "    default:" +
+            "  }");
+    compileSnippet("void",
+        "for (int i : new int[] {1, 2, 3, 4}) " +
+            "  switch(i) { " +
+            "    case 1:" +
+            "    case 2:" +
+            "    default:" +
+            "  }");
+    compileSnippet("void",
+        "switch (4) {" +
+            "case 1:" +
+            "  switch (2) {" +
+            "    case 2:" +
+            "    case 4:" +
+            "    default:" +
+            "  }" +
+            "}");
+    compileSnippet("void",
+        "if (true == false) " +
+            "  switch (7) {" +
+            "    case 4: {" +
+            "      break;" +
+            "    }" +
+            "  }" +
+            "else " +
+            "  switch (8) {" +
+            "    case 9:" +
+            "  }");
+
+    compileSnippet("void",
+        "while(true)" +
+            "  switch(99) { " +
+            "    default:" +
+            "  }");
+
+    compileSnippet("void",
+        "do" +
+            "  switch(0) { " +
+            "    default:" +
+            "  }" +
+            "while (false);");
+
+    compileSnippet("void",
+        "foo:" +
+            "  switch(123) { " +
+            "    default:" +
+            "  }");
+  }
+
   @Override
   protected void optimizeJava() {
   }

